### PR TITLE
Stub objects for Bookmarks sync

### PR DESCRIPTION
**Pull request by bwaresiak**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**: 

Task/Issue URL: https://app.asana.com/0/0/1206831058661531/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2593
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2418
What kind of version bump will this require?: Major

**Description**:

Provide implementation for stub objects.

**Steps to test this PR**:

On platforms, navigate to debug menu and populate stubs in the DB. UX should not be affected by stubs.

Validate tests if they cover all scenarios.

Smoke tests sync.

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift [hunk 1]
```diff
@@ -815,6 +815,128 @@ internal class BookmarksProviderTests: BookmarksProviderTestsBase {
...
+            newBookmark.uuid = "3"
+            try! context.save()
+        }
+
+        let sent = try await provider.fetchChangedObjects(encryptedUsing: crypter)
+
+        // Only Root and "3" should be sent
+        XCTAssertEqual(sent.count, 2)
+
+        let sentRootData = sent.first(where: { $0.payload["id"] as? String == rootFolder.uuid })
```
Comment (ayoy) [inline Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift:855]: We could use SyncableBookmarkAdapter to make this code a bit simpler:

```swift
XCTAssertEqual(sent.map({ SyncableBookmarkAdapter(syncable: $0).uuid }), [rootFolder.uuid, "3"])

let sentRootData = sent.first(where: { SyncableBookmarkAdapter(syncable: $0).uuid == rootFolder.uuid })
```

But then again, SyncableBookmarkAdapter doesn't support folder changes at the moment, so the rest needs to stay as is.

File: Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift [hunk 2]
```diff
@@ -815,6 +815,128 @@ internal class BookmarksProviderTests: BookmarksProviderTestsBase {
...
+        let bookmarkTree = BookmarkTree {}
+
+        context.performAndWait {
+            BookmarkUtils.prepareFoldersStructure(in: context)
+            bookmarkTree.createEntities(in: context)
+            try! context.save()
+        }
+
+        let received: [Syncable] = [
+            .rootFolder(children: ["1", "2", "3"]), // Creates a Stub with id 2
```
Comment (ayoy) [inline Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift:879]: [...] with ids 2 and 3

File: Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift [hunk 3]
```diff
@@ -815,6 +815,128 @@ internal class BookmarksProviderTests: BookmarksProviderTestsBase {
...
+        var rootFolder = try await handleInitialSyncResponse(received: received, clientTimestamp: Date(), serverTimestamp: "1234", in: context)
+        assertEquivalent(withTimestamps: false, rootFolder, BookmarkTree(lastChildrenArrayReceivedFromSync: ["1", "2", "3"]) {
+            Bookmark(id: "1")
+            Bookmark(id: "2", isStub: true)
+            Bookmark(id: "3", isStub: true)
+        })
+
+        // Simulate two kinds of "removal":
+        // - Removal only from children list.
+        // - Removal of entity and from children list.
```
Comment (ayoy) [inline Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift:892]: Could we please adjust the comment to make it more clear?

```suggestion
        // - "2" is only removed from children list.
        // - "3" is removed from children list and deleted.
```

File: Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift [hunk 4]
```diff
@@ -815,6 +815,128 @@ internal class BookmarksProviderTests: BookmarksProviderTestsBase {
...
+        var rootFolder = try await handleInitialSyncResponse(received: received, clientTimestamp: Date(), serverTimestamp: "1234", in: context)
+        assertEquivalent(withTimestamps: false, rootFolder, BookmarkTree(lastChildrenArrayReceivedFromSync: ["1"]) {
+            Folder(id: "1", lastChildrenArrayReceivedFromSync: ["2"]) {
+                Bookmark(id: "2", isStub: true)
+            }
+        })
+
+        // Simulate two kinds of "removal":
+        // - Removal only from children list.
+        // - Removal of entity and from children list.
```
Comment (ayoy) [inline Tests/SyncDataProvidersTests/Bookmarks/BookmarksProviderTests.swift:929]: I think this comment does not apply here, we could just remove it as the test name is pretty clear.

**Review by ayoy — APPROVED** [review]
All works great! I've left a few minor comments, mostly to make tests more clear. Thanks @bwaresiak!

File: Sources/SyncDataProviders/Bookmarks/internal/BookmarksResponseHandler.swift [hunk 5]
```diff
@@ -171,11 +188,10 @@ final class BookmarksResponseHandler {
             var queue = queues.removeFirst()
             let parentUUID = parentUUIDs.removeFirst()
             let parent = BookmarkEntity.fetchFolder(withUUID: parentUUID, in: context)
-            assert(parent != nil)
 
             // For non-first sync we rely fully on the server response
             if !shouldDeduplicateEntities {
-                parent?.childrenArray.forEach { parent?.removeFromChildren($0) }
+                (parent?.children?.array as? [BookmarkEntity] ?? []).forEach { parent?.removeFromChildren($0) }
```
[the remark this answers is not in the document]
Comment (bwaresiak) [reply Sources/SyncDataProviders/Bookmarks/internal/BookmarksResponseHandler.swift:194]: If bookmark gets undeleted it should still get processed here correctly. Problem was that the stubs were not cleaned up properly as they were not removed from parent here.

File: Sources/SyncDataProviders/Bookmarks/internal/SyncableBookmarkAdapter.swift [hunk 6]
```diff
@@ -92,7 +92,11 @@ extension Syncable {
                     if BookmarkEntity.Constants.favoriteFoldersIDs.contains(uuid) {
                         return bookmark.favoritesArray.compactMap(\.uuid)
                     }
-                    return bookmark.childrenArray.compactMap(\.uuid)
+                    let validChildrenIds = bookmark.childrenArray.compactMap(\.uuid)
+
+                    // Take stubs into account - we don't want to remove them.
+                    let stubIds = (bookmark.children?.array as? [BookmarkEntity] ?? []).filter({ $0.isStub }).compactMap(\.uuid)
+                    return validChildrenIds + stubIds
```
[the remark this answers is not in the document]
Comment (bwaresiak) [reply Sources/SyncDataProviders/Bookmarks/internal/SyncableBookmarkAdapter.swift:99]: Yup, my bad, fixed and added a test. 👍

**Review by ayoy — APPROVED** [review]
LGTM! Thanks for addressing the feedback!

**Pull request merged**